### PR TITLE
IDEMPIERE-5886 - Add Foreign Key Constraint Error Message to AD_Column

### PR DESCRIPTION
**Pull request by PeterTakacs300**
JIRA ticket: https://idempiere.atlassian.net/browse/IDEMPIERE-5886

# Pull Request Checklist

- [x] My code follows the [code guidelines](https://wiki.idempiere.org/en/Contributing_to_Trunk) of this project
- [x] My code follows the best practices of this project
- [x] I have performed a self-review of my own code
- [x] My code is easy to understand and review. 
- [x] In hard-to-understand areas, I have commented my code.
- [x] My changes generate no new warnings
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added unit tests that prove my fix is effective or that my feature works
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings
- [ ] I have made corresponding changes to the documentation as follows:
  - [ ] New feature (non-breaking change which adds functionality): I have created the New Feature page in the project wiki explaining the functionality and how to use it. If relevant, I have committed sample data to the core seed to have usable examples in GardenWorld.
  - [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected): I have documented the change in a clear way that everyone in the community can understand the impact of the change.
  - [ ] Improvement (improves and existing functionality): This documentation is needed if the improvement changes the way the user interacts with the system or the outcome of a process/task changes. If it is just, for instance, a performance improvement, documentation might not be needed. 
- [ ] The changed/added documentation is in the project wiki (not privately-hosted pdf files or links pointing to a company website) and is complete and self-explanatory.

### CODE REVIEW

**Comment by PeterTakacs300** [comment]
Note: the getNameOfChildRecordFoundError method is implemented only for postgres - I'm not sure what the error message looks like for oracle.

File: org.adempiere.base/src/org/compiere/db/AdempiereDatabase.java [hunk 1]
```diff
@@ -333,6 +333,26 @@ public Connection getDriverConnection (String dbUrl, String dbUid, String dbPwd)
...
+     * ERROR: update or delete on table "m_product_category" 
+     * violates foreign key constraint "mprodcategory_mdiscountsbreak" 
+     * on table "m_discountschemabreak" 
+     * Detail: Key (m_product_category_id)=(50000) is still referenced from table "m_discountschemabreak".
+     * </p>
+     * 
+     * @param e
+     * @return constraint name
+     */
+	public String getNameOfChildRecordFoundError(Exception e);
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/db/AdempiereDatabase.java:355]: lets make this obvious and name it as `getForeignKeyConstraint(Exception e)` instead

File: org.adempiere.base/src/org/compiere/db/AdempiereDatabase.java [hunk 2]
```diff
@@ -333,6 +333,26 @@ public Connection getDriverConnection (String dbUrl, String dbUid, String dbPwd)
 	public boolean forUpdate(PO po, int timeout);
 	
 	public String getNameOfUniqueConstraintError(Exception e);
+	
+	/**
+     * <p>
+     * The "child record found error" contains the 
+     * unique constraint name after the second occurrence
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/db/AdempiereDatabase.java:340]: this is foreign key constraint, not unique constraint.

File: org.compiere.db.oracle.provider/src/org/compiere/db/DB_Oracle.java [hunk 3]
```diff
@@ -950,6 +950,15 @@ public String getNameOfUniqueConstraintError(Exception e) {
 		return info.substring(fromIndex + 1, toIndex);
 	}
 
+	@Override
+	public String getNameOfChildRecordFoundError(Exception e) {
+		String info = e.getMessage();
+		
+		// TODO - to be implemented...
+		
+		return info;
```
Comment (hengsin) [inline org.compiere.db.oracle.provider/src/org/compiere/db/DB_Oracle.java:959]: this is bad practise, should just return empty string or null.

**Comment by PeterTakacs300** [comment]
Hi @hengsin, changes committed.

**Comment by hengsin** [comment]
hi @CarlosRuiz-globalqss , ok to proceed without implementation for Oracle ?

File: org.compiere.db.oracle.provider/src/org/compiere/db/DB_Oracle.java [hunk 4]
```diff
@@ -950,6 +950,12 @@ public String getNameOfUniqueConstraintError(Exception e) {
 		return info.substring(fromIndex + 1, toIndex);
 	}
 
+	@Override
+	public String getForeignKeyConstraint(Exception e) {
```
Comment (CarlosRuiz-globalqss) [inline org.compiere.db.oracle.provider/src/org/compiere/db/DB_Oracle.java:954]: Please apply the following patch:
[pr2069PR.patch.txt](https://github.com/idempiere/idempiere/files/13114374/pr2069PR.patch.txt)

It implements oracle, and fixes error when postgresql is in German (for foreign and for unique discovery).

I also implemented a generic more explanatory message stating the table and column involved in the foreign key error.

Also, moved the code from PO to DBException, and added methods on CLogger to report the friendly message to user still saving the exception (there are plugins that uses the exception for other purposes).

**Review by CarlosRuiz-globalqss — CHANGES_REQUESTED** [review]

File: org.compiere.db.oracle.provider/src/org/compiere/db/DB_Oracle.java (hunk 4 above)
Comment (PeterTakacs300) [reply org.compiere.db.oracle.provider/src/org/compiere/db/DB_Oracle.java:954]: thanks Carlos, applied the patch, passed my basic tests

**Review by CarlosRuiz-globalqss — APPROVED** [review]

**Pull request merged**